drivers: spi: siwx91x: SPI driver fixes to enable higher burst rates

smalae opened the pull request:
1. Setting INTF PLL to 160 MHz so GSPI has a stable, high-rate base clock, and QSPI gets the expected 80 MHz. 
2. Rounding the GSPI frequency down to a valid step derived from the INTF PLL (f = PLL/(2*div)), and logging the requested vs programmed rate. 
3. Enabling higher GPDMA burst sizes (1/2/4) when safe (SPI ≥ 10 MHz and both channels use GPDMA). The burst size is selected from the largest value that satisfies TX/RX buffer alignment, and FIFO thresholds are set accordingly. A temporary limitation notes DMA flow-control issues at low SPI rates, so bursts are gated behind the 10 MHz threshold. 
4. Align mosi_overrun to 32 bytes
5. Fixing 16-bit transfers by swapping write/read data paths in the GSPI controller (GSPI_CONFIG2_b), ensuring the correct signal presentation on the SPI lines. 
6. Guarding async callback registration to avoid build errors when using blocking DMA transfers without CONFIG_SPI_ASYNC. 
7. Updating tests: increase CONFIG_SPI_IDEAL_TRANSFER_DURATION_SCALING from 24 → 26 to account for the (slightly) increased transfer time due to burst-size calculation.
8.  Enable the spi_loopback test for the PSRAM board siwx917_rb4342a.

smalae committed:
drivers: clock_control: siwx91x: Set INTF_PLL_FREQUENCY to 160 MHz

The GSPI and QSPI peripherals run on the interface PLL clock.
To ensure correct operation, the interface PLL frequency should
be set to 160 MHz. This provides a base clock of 80 MHz to the
QSPI peripheral, which matches its required operating frequency.
The GSPI peripheral can continue to operate at higher frequencies
as it receives the full interface PLL clock.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: spi: siwx91x: Add helper to round GSPI freq to valid range

Add gspi_siwx91x_pick_lower_freq() to select the nearest lower
supported GSPI clock frequency based on the requested value.
The GSPI clock can operate only at discrete frequencies such as
80 MHz, 40 MHz, 26.6 MHz, 20 MHz, and so on (80 MHz divided by
integer factors). If the requested frequency does not match one
of these valid steps, it is rounded down to the nearest lower
supported frequency.

The driver now logs both the requested and the actual programmed
frequency along with the divider value, helping users verify the
effective SPI clock configuration.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: spi: siwx91x: Add support for higher burst rates using GPDMA

Enable higher SPI burst sizes when using GPDMA to improve throughput.
Bursts are only enabled when both TX and RX channels use GPDMA and
the SPI clock frequency is at least 10 MHz, as DMA flow control is
unreliable at lower SPI rates.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: spi: siwx91x: Swap write and read data inside GSPI controller

Swap the write and read data paths within the GSPI controller to ensure
correct data is seen on the SPI lines during 16-bit transfers.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
drivers: spi: siwx91x: Guard callback registration

Guard callback registration with CONFIG_SPI_ASYNC to avoid
build error for blocking SPI transfers using DMA

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
tests: drivers: spi: siwx91x: Adjust transfer scaling factor

Increase CONFIG_SPI_IDEAL_TRANSFER_DURATION_SCALING to
compensate for the additional burst size calculation in
the SPI driver, which slightly increases transfer time.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

jerome-pouiller commented on tests/drivers/spi/spi_loopback/testcase.yaml:
```diff
@@ -360,6 +360,12 @@ tests:
+      - siwx917_rb4342a
+    extra_configs:
+      - CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA=y
+      - CONFIG_SPI_IDEAL_TRANSFER_DURATION_SCALING=26
```
Redundant with the `.conf` file, no?

smalae committed:
drivers: spi: siwx91x: Align mosi_overrun to 32 bytes

Updated mosi_overrun alignment from 4 bytes to 32 bytes to
ensure compatibility with the GPDMA driver. This prevents
alignment-related errors during DMA transfers.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae committed:
tests: drivers: spi: siwx91x: Enable spi_loopback test for PSRAM board

Add overlay and config files for siwx917_rb4342a
to enable the spi_loopback test.

Signed-off-by: Sai Santhosh Malae <Santhosh.Malae@silabs.com>

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -76,13 +76,35 @@ static bool spi_siwx91x_is_dma_enabled_instance(const struct device *dev)
+				  uint32_t *div_out)
+{
+	/* Calculate divider that ensures freq <= requested */
+	uint32_t divider = (clock_hz + (2U * requested_hz) - 1U) / (2U * requested_hz);
```
Updated

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -76,13 +76,35 @@ static bool spi_siwx91x_is_dma_enabled_instance(const struct device *dev)
+	/* Compute the actual achievable frequency */
+	uint32_t actual_hz = clock_hz / (2U * divider);
+
+	if (divider == 0U) {
```
Done

smalae commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -423,6 +428,52 @@ static size_t gspi_siwx91x_longest_transfer_size(struct spi_context *instance_ct
+	}
+
+	return 1;
+}
```
Updated as per the suggestion with minor adjustments.

smalae commented on tests/drivers/spi/spi_loopback/testcase.yaml:
```diff
@@ -360,6 +360,12 @@ tests:
+      - CONFIG_SPI_SILABS_SIWX91X_GSPI_DMA=y
+      - CONFIG_SPI_IDEAL_TRANSFER_DURATION_SCALING=26
+    extra_args:
+      - DTC_OVERLAY_FILE="boards/siwx917_rb4338a.overlay"
```
Done

jhedberg commented on drivers/spi/spi_silabs_siwx91x_gspi.c:
```diff
@@ -48,7 +49,7 @@ struct gspi_siwx91x_config {
 	clock_control_subsys_t clock_subsys;
 	const struct pinctrl_dev_config *pcfg;
-	uint8_t mosi_overrun __aligned(4);
+	uint8_t mosi_overrun __aligned(32);
```
@tbursztyka are you ok with this explanation?
